### PR TITLE
Documentation 📚 

### LINKED ISSUE #7: Documentation and/or changelog?
I am very confused on how to migrate my chatbox programs. Simply upgrading version didn't work, and downgrading again to the old version and changing endpoint didn't work either.

### PR DESCRIPTION
**Pull request by piprett**
I have made some documentation. It is in no way perfect, and I am in no way good at English. Please do nitpick.
Fixes #7.
I used obsidian.md while writing this, so my markdown may be a bit off.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,20 +1,229 @@
 # SwitchChat v3
-### SwitchCraft Chatbox v2 Module for Node.js
 
-# LOOKING FOR MAINTAINERS
+SwitchCraft Chatbox v2 module for Node.js written in TypeScript
 
-[![npm version](https://badge.fury.io/js/switchchat.svg)](https://www.npmjs.org/package/switchchat)
```
Comment (Ale32bit) [inline README.md:4]: Any reason behind removing the npm version badge?

**Review by Ale32bit — APPROVED** [review]

File: README.md (hunk 1 above)
Comment (piprett) [reply README.md:4]: image
Sorry? I can't see this

File: README.md [hunk 2]
```diff
@@ -1,18 +1,229 @@
...
+
+- user - `User` that goes AFK
+
+### Return from AFK
+
+- user - `User` that goes AFK
+
+### Ingame Chat
+
+- text - `String` of the contents of the chat message
```
Comment (Lemmmy) [inline README.md]: JavaScript primitives are written in lowercase, so this should be `string` (and in all other instances of this)

File: README.md [hunk 3]
```diff
@@ -1,18 +1,229 @@
...
 
-```js
-// work in progress
+// Create the client
+const sc = new Client('YOUR-CHATBOX-TOKEN');
+
+// Add event listener for when a command is fired
+sc.on('command', (cmd) => {
+  // Only respond if the command is `\helloworld`
+  if (cmd.command == 'helloworld') {
```
Comment (Lemmmy) [inline README.md]: Use [triple equals](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Strict_equality) here (`===`)

```suggestion
  if (cmd.command === 'helloworld') {
```

File: README.md [hunk 4]
```diff
@@ -1,18 +1,229 @@
...
+| Restart scheduled | server_restart_scheduled | When a restart gets scheduled                |
+| Restart cancelled | server_restart_cancelled | When a restart gets cancelled                |
+
+### AFK
+
+- user - `User` that goes AFK
+
+### Return from AFK
+
+- user - `User` that goes AFK
```
Comment (Lemmmy) [inline README.md]:
```suggestion
### AFK return

- user - `User` that went AFK
```

File: README.md [hunk 5]
```diff
@@ -1,18 +1,229 @@
...
+| Return from AFK   | afk_return               | When someone returnes from AFK               |
+| Ingame Chat       | chat_ingame              | When someone sends an ingame chat message    |
+| Discord Chat      | chat_discord             | When someone sends a chat message on Discord |
+| Chatbox Broadcast | chat_chatbox             | When another chatbox broadcasts              |
+| Chatbox command   | command                  | When someone sends a chatbox command         |
+| Death             | death                    | When someone dies ingame                     |
+| Join              | join                     | When someone joines the game                 |
+| Disconnect        | leave                    | When someone leaves the game                 |
+| Restart scheduled | server_restart_scheduled | When a restart gets scheduled                |
+| Restart cancelled | server_restart_cancelled | When a restart gets cancelled                |
```
Comment (Lemmmy) [inline README.md]:
```suggestion
| Friendly name     | Name                     | Trigger                                    |
| ----------------- | ------------------------ | ------------------------------------------ |
| AFK               | afk                      | A player went AFK                          |
| AFK return        | afk_return               | A player returned from being AFK           |
| In-game chat      | chat_ingame              | A player sent a message in-game            |
| Discord chat      | chat_discord             | A player sent a message from Discord       |
| Chatbox message   | chat_chatbox             | A chatbox sent a public message            |
| Chatbox command   | command                  | A player sent a chatbox command in-game    |
| Death             | death                    | A player died                              |
| Join              | join                     | A player joined the server                 |
| Disconnect        | leave                    | A player left the server                   |
| Restart scheduled | server_restart_scheduled | The server will restart soon               |
| Restart cancelled | server_restart_cancelled | The scheduled server restart was cancelled |
```

File: README.md [hunk 6]
```diff
@@ -1,18 +1,229 @@
...
+- args - Array of `string`s that contain the args that come after the command
+- ownerOnly - `Boolean` that says if this command was visible to everyone's chatbox, or only the owner
+
+### Death
+
+- text - `String` of the death message
+- rawText - Raw `string` of the death message
+- renderedText - `RenderedTextObject` of the death message
+- user - `User` that died
+- source - `User` | `null` that killed the user. Only sent if they was killed by a player
```
Comment (Lemmmy) [inline README.md]:
```suggestion
- source - `User` | `null` that killed the user. Only set if they were killed by another player
```

File: README.md [hunk 7]
```diff
@@ -1,18 +1,229 @@
...
+
+- user - `User` that sent the chatbox command
+- command - The first part of the command they sent. Eg. \\**command** arg1 arg2
+- args - Array of `string`s that contain the args that come after the command
+- ownerOnly - `Boolean` that says if this command was visible to everyone's chatbox, or only the owner
+
+### Death
+
+- text - `String` of the death message
+- rawText - Raw `string` of the death message
```
Comment (Lemmmy) [inline README.md]:
```suggestion
- text - `string` of the death message contents, without formatting codes
- rawText - `string` of the death message contents, with its original formatting codes
```

File: README.md [hunk 8]
```diff
@@ -1,18 +1,229 @@
...
+- edited - `Boolean` if the message has been edited
+
+### Chatbox Broadcast
+
+- text - `String` of the contents of the broadcast
+- rawText - `String` of the raw contents of the broadcast
+- renderedText - `RenderedTextObject` of the rendered text of the broadcast
+- user - `User` of the owner of the chatbox that broadcasted
+- name - `String` name of the chatbox of the chatbox that broadcasted
+- rawName - Raw `string` name of the chatbox that broadcasted
```
Comment (Lemmmy) [inline README.md]:
```suggestion
### Chatbox message

- text - `string` of the contents of the message, without formatting codes
- rawText - `string` of the raw contents of the message, with its original formatting codes
- renderedText - `RenderedTextObject` of the rendered text of the message
- user - `User` of the owner of the chatbox
- name - `string` name of the chatbox, without formatting codes
- rawName - `string` name of the chatbox, with its original formatting codes
```

File: README.md [hunk 9]
```diff
@@ -1,18 +1,229 @@
...
+- user - `User` that sent the message
+
+### Discord Chat
+
+- text - `String` of the contents of the chat message
+- rawText - `String` of the raw contents of the chat message
+- renderedText - `RenderedTextObject` of the rendered text of the chat message
+- discordId - `String` of the Discord snowflake ID of the message
+- discordUser - `DiscordUser` that sent the message
+- edited - `Boolean` if the message has been edited
```
Comment (Lemmmy) [inline README.md]:
```suggestion
### Discord chat

- text - `string` of the contents of the chat message, without formatting codes
- rawText - `string` of the raw contents of the chat message, with its original formatting codes
- renderedText - `RenderedTextObject` of the rendered text of the chat message
- discordId - `string` of the Discord snowflake ID of the message
- discordUser - `DiscordUser` that sent the message
- edited - `boolean` if the message has been edited
```

File: README.md [hunk 10]
```diff
@@ -1,18 +1,229 @@
...
+- user - `User` of the owner of the chatbox that broadcasted
+- name - `String` name of the chatbox of the chatbox that broadcasted
+- rawName - Raw `string` name of the chatbox that broadcasted
+
+### Chatbox Command
+
+- user - `User` that sent the chatbox command
+- command - The first part of the command they sent. Eg. \\**command** arg1 arg2
+- args - Array of `string`s that contain the args that come after the command
+- ownerOnly - `Boolean` that says if this command was visible to everyone's chatbox, or only the owner
```
Comment (Lemmmy) [inline README.md]:
```suggestion
### Chatbox command

- user - `User` that sent the chatbox command
- command - The first part of the command they sent. Eg. \\**command** arg1 arg2
- args - Array of `string`s that contain the args that come after the command
- ownerOnly - `boolean` if the command is an owner-only command (`^command`)
```

File: README.md [hunk 11]
```diff
@@ -1,18 +1,229 @@
...
+
+- user - `User` that left the server
+
+### Server restart scheduled
+
+Not fully implemented
+
+### Server restart cancelled
+
+Not fully implemented
```
Comment (Lemmmy) [inline README.md:126]: Heads-up: with #11 these are implemented

File: README.md [hunk 12]
```diff
@@ -1,18 +1,229 @@
...
+
+## Server -> Client packets
+
+You are also able to listen after these packets that are sent by SwitchCraft to the client.
+
+| Friendly name     | Name    | Trigger                                                   |
+| ----------------- | ------- | --------------------------------------------------------- |
+| Hello packet      | ready   | When switchchat has successfully connected to SwitchCraft |
+| Playerlist packet | players | When you connect, and when a player joines/leaves         |
```
Comment (Lemmmy) [inline README.md]:
```suggestion
| Friendly name     | Name    | Trigger                                                   |
| ----------------- | ------- | --------------------------------------------------------- |
| Hello packet      | ready   | When SwitchChat has successfully connected to SwitchCraft |
| Playerlist packet | players | When you connect, and when a player joins/leaves          |
```

File: README.md [hunk 13]
```diff
@@ -1,18 +1,229 @@
...
+- players - Array of `User`s.
+
+### Error packet
+
+- error - `String` that tells you what type of error that occoured. You can find possible errors at the [SwitchCraft wiki](https://docs.sc3.io/chatbox/websocket.html#error-packet)
+- text - `String` that is a human readable version of `error`.
+
+### Closing packet
+
+- closeReason - `String` reson for closing the connection. You can find possible values at the [SwitchCraft wiki](https://docs.sc3.io/chatbox/websocket.html#closing-packet)
```
Comment (Lemmmy) [inline README.md]:
```suggestion
- error - `String` that tells you what type of error that occoured. You can find possible errors at the [SwitchCraft documentation](https://docs.sc3.io/chatbox/websocket.html#error-packet).
- text - `String` that is a human readable version of `error`.

### Closing packet

- closeReason - `String` reson for closing the connection. You can find possible values at the [SwitchCraft documentation](https://docs.sc3.io/chatbox/websocket.html#closing-packet).
```

File: README.md [hunk 14]
```diff
@@ -1,18 +1,229 @@
...
+- pronouns - Not yet implemented in SwitchChat v3
+- world - The world the player is in. Either the full name like `minecraft:nether` or null
+- afk - If the player is AFK
+
+### Discord User
+
+- id - The discord snowflake ID of the user
+- name - The discord name of the user
+- displayName - The displayname of the user
+- discriminator - The four numbers that identify this user from all the other with the same name
```
Comment (Lemmmy) [inline README.md]:
```suggestion
- id - The Discord snowflake ID of the user
- name - The Discord name of the user
- displayName - The user's server nickname on Discord, or their username if it is not set
- discriminator - The user's discriminator on Discord
```

File: README.md [hunk 15]
```diff
@@ -1,18 +1,229 @@
...
+- avatar - URL to the avatar of the user
+- roles - Array consisting of roles.
+
+### Rendered Text Object
+
+Weird JSON object that minecraft uses for rendering.
```
Comment (Lemmmy) [inline README.md]:
```suggestion
Minecraft's serialised [raw JSON text format](https://minecraft.fandom.com/wiki/Raw_JSON_text_format). See the [SwitchCraft documentation](https://docs.sc3.io/chatbox/websocket.html#raw-json-text-object) for more information on how this is used.
```

File: README.md [hunk 16]
```diff
@@ -1,18 +1,229 @@
...
+
+### Mode
+
+- markdown - To use markdown as formatting for outgoing chatbox messages
+- format - To use minecraft formatting for outgoing chatbox messages
```
Comment (Lemmmy) [inline README.md]:
```suggestion
The mode to use for outgoing chatbox messages (`.say()`, `.tell()`). 

- markdown - Discord-like [Markdown syntax](https://support.discord.com/hc/en-us/articles/210298617-Markdown-Text-101-Chat-Formatting-Bold-Italic-Underline-). Supports URLs, but not colours.
- format - Minecraft-like [formatting codes](https://minecraft.fandom.com/wiki/Formatting_codes) using ampersands (e.g. `&e` for yellow). Supports colours, but not URLs.
```

**Review by Lemmmy — CHANGES_REQUESTED** [review]
Thank you for your work on this! My review leaves a lot of small nitpicks/nags, mostly regarding formatting, grammar, and consistency with [docs.sc3.io](https://docs.sc3.io/chatbox/websocket.html). I will probably do another pass over this after it is merged, but what you have here is a fantastic start.

When #11 is merged and deployed, I will also add links between the various places of documentation (README, sc-docs, etc.), but don't worry about that in your PR.

**Comment by Lemmmy** [comment]
As outlined in #11, I have created a stub page for documentation at [docs.sc3.io](https://docs.sc3.io/chatbox/switchchat.html) which we can work on after this PR is merged. The tsdocs are also available [online](https://docs.sc3.io/library/switchchat/).

File: README.md (hunk 11 above)
Comment (piprett) [reply README.md:126]: Would you like me to document the code changes in #11, or will you take care of this?

Comment (Lemmmy) [reply README.md:126]: I can take care of it if you prefer.

**Review by Lemmmy — APPROVED** [review]
LGTM, thanks for your work on this!

**Pull request merged**